lib/src/util/dart_type_utilities.dart

```diff
@@ -49,6 +49,69 @@ class DartTypeUtilities {
     return null;
   }
 
+  /// Return whether the canonical elements of two elements are equal.
+  static bool canonicalElementsAreEqual(Element element1, Element element2) =>
+      getCanonicalElement(element1) == getCanonicalElement(element2);
+
+  /// Returns whether the canonical elements from two nodes are equal.
+  ///
+  /// As in, [getCanonicalElementFromIdentifier], the two nodes must be
+  /// [Expression]s in order to be compared (otherwise `false` is returned).
+  ///
+  /// The two nodes must both be a [SimpleIdentifier], [PrefixedIdentifier], or
+  /// [PropertyAccess] (otherwise `false` is returned).
+  ///
+  /// If the two nodes are PrefixedIdentifiers, or PropertyAccess nodes, then
+  /// `true` is returned only if their canonical elements are equal, in
+  /// addition to their prefixes' and targets' (respectfully) canonical
+  /// elements.
+  ///
+  /// There is an inherent assumption about pure getters. For example:
```

Comment:
Thanks for calling this out! I'm not sure that this is a good assumption to make, but given that it's explicit I'm willing to leave it.

Reply:
Aside: I'm intrigued by the idea of `@sideEffects`-related annotations and what they could enable us to do with analysis.  I see there's a conversation started elsewhere so I'll follow with interest!

Reply:
I do think it's an incorrect assumption. Getters are not guaranteed to be side-effect free, so we should not assume that they are. Making that assumption leads to false positives, which is a problem in any kind of static analysis.

That said, I'm assuming that the assumption isn't new, and at least now all of the logic based on this assumption is in one place. That's definitely an improvement.

Reply:
I filed an issue for `@sideEffects`, https://github.com/dart-lang/sdk/issues/35513.